feat: add send screen amount field max balance error

lujakob opened the pull request:
### Describe the changes you have made in this PR

Add error color to field border and label to inform user about selected amount exceeds balance. 

### Link this PR to an issue [optional]

Fixes #2415

### Type of change

(Remove other not matching type)

- `feat`: New feature (non-breaking change which adds functionality)

### Screenshots of the changes [optional]

<img width="383" alt="Bildschirmfoto 2023-05-23 um 16 42 42" src="https://github.com/getAlby/lightning-browser-extension/assets/886152/63596175-9981-4389-9d1b-d528003b88da">

### How has this been tested?

Manually

### Checklist

- [x] My code follows the style guidelines of this project and performed a self-review of my own code
- [x] New and existing tests pass locally with my changes
- [x] I checked if I need to make corresponding changes to the documentation (and made those changes if needed)

lujakob commented:
@reneaaron I only made changes in the LNURLPay page. Does it make sense to add it to the Keysend page as well? Any more I am missing?

lujakob commented:
@reneaaron I noticed the max value of the range mentioned and shown in the screenshot above is fixed. I'll adapt it to be either the current fixed or `auth.account.balance`, because it doesn't it is contrarian to display a range higher than the max, agree?

` const amountMax = Math.min(Math.floor(+details.maxSendable / 1000), auth?.account?.balance || 0);`

reneaaron commented:
Very nice, thanks for the PR! :muscle: 

> @reneaaron I noticed the max value of the range mentioned and shown in the screenshot above is fixed. I'll adapt it to be either the current fixed or auth.account.balance, because it doesn't it is contrarian to display a range higher than the max, agree?

I think the intention of this label is to show the range the receiver is able to accept and would probably try to keep that separate from the balance display (which is also marked red in case of an error). 

I currently can't think of anything else than the Keysend screen, no. (other screens like the receive screen are not limited by your balance)

lujakob commented:
Thanks for checking.

> Very nice, thanks for the PR! 💪
> 
> > @reneaaron I noticed the max value of the range mentioned and shown in the screenshot above is fixed. I'll adapt it to be either the current fixed or auth.account.balance, because it doesn't it is contrarian to display a range higher than the max, agree?
> 
> I think the intention of this label is to show the range the receiver is able to accept and would probably try to keep that separate from the balance display (which is also marked red in case of an error).
> 

I reverted this change.

> I currently can't think of anything else than the Keysend screen, no. (other screens like the receive screen are not limited by your balance)

I added the new changes to keysend. What I noticed, there is not `hint` about the amount underneath left side. So it's not really obvious, why the input field suddenly has a red border. I suppose it's just missing and added it. See my two screens before and after. I key keysend page should not have a balance hint, I can remove it. But then I think it needs to be improved somehow else...

<img width="1019" alt="Bildschirmfoto 2023-05-24 um 13 20 02" src="https://github.com/getAlby/lightning-browser-extension/assets/886152/3de157a9-647f-4989-b789-2dd6780be5b4">

<img width="995" alt="Bildschirmfoto 2023-05-24 um 13 22 01" src="https://github.com/getAlby/lightning-browser-extension/assets/886152/8a1e06d5-39d5-4771-904b-9fa2e16d20a3">

reneaaron commented:
@lujakob Including the balance here absolutely makes sense, thanks for adding it! I'll do some final testing later today :+1:

reneaaron commented on src/app/components/form/DualCurrencyField/index.tsx:
```diff
@@ -9,6 +10,8 @@ export type Props = {
   label: string;
   hint?: string;
+  amountExceeded?: boolean;
+  maxExceeded?: boolean;
```
Since there is also a minimum value for the range it would probably make sense to change this to `rangeExceeded` and also handle amounts < min for consistency?

reneaaron commented on src/app/screens/LNURLPay/index.tsx:
```diff
@@ -518,7 +524,7 @@ function LNURLPay() {
                         label={tCommon("actions.confirm")}
                         loading={loadingConfirm}
-                        disabled={loadingConfirm || !valueSat}
+                        disabled={loadingConfirm || !valueSat || amountExceeded}
```
Should the `rangeExceeded` be included in here as well?

lujakob replied:
Updated.

lujakob commented:
@reneaaron I updated both change requests and adapted the changes for both Lnurlpay and Keysend.

reneaaron reviewed (commented):
Isn't it a bit strange that every consumer of this component has to implement those checks again? 

Would it be more consistent to add an optional field for the current balance to the `DualCurrencyField` component and check for those values in there?

lujakob commented:
> Isn't it a bit strange that every consumer of this component has to implement those checks again?
> 
> Would it be more consistent to add an optional field for the current balance to the `DualCurrencyField` component and check for those values in there?

I'm not really sure what you mean. Passing the `auth?.account?.balance` to the `DualCurrencyField` instead of calculating `const amountExceeded = +amountSat > (auth?.account?.balance || 0);` in the parent and passing `amountExceeded` into `DualCurrencyField`?

I think a problem is we need `amountExceeded` in the parent as well to disable the button.

rolznz commented on src/app/components/form/DualCurrencyField/index.tsx:
```diff
@@ -85,8 +91,13 @@ export default function DualCurrencyField({
+          <span
+            className={classNames(
+              "text-xs text-gray-700 dark:text-neutral-400",
+              (!!min || !!max) && !!rangeExceeded && "text-red-500"
```
Why is `(!!min || !!max)` check needed here? it is already checked in the enclosing scope (line 93)

rolznz replied:
You shouldn't need to do `!!` on a boolean? values

lujakob replied:
The reason why I had `(!!min || !!max)` here is, in Keysend component I have `const rangeExceeded = +amountSat < amountMin;` - `amountMin` is `1` and if for some reason `amountSat` would be `undefined` I would end up with `+undefined < 1` which is `false`. This cased a small amount of uncertainty and I wanted to make sure this rangeExceeded is only being applied when there actually is a min or max value. But probably it is not needed.

reneaaron reviewed (approved):
tACK

rolznz commented:
@reneaaron sorry, I mean it doesn't accept `undefined` as false. Some other classnames libraries like https://www.npmjs.com/package/clsx do

rolznz commented on src/app/components/form/DualCurrencyField/index.tsx:
```diff
@@ -125,7 +138,12 @@ export default function DualCurrencyField({
+        <p
+          className={classNames(
+            "my-1 text-xs text-gray-700 dark:text-neutral-400",
+            !!amountExceeded && "text-red-500 dark:text-red-500"
```
why is the dark: version the same? can't it be omitted? (multiple places in this file)

reneaaron replied:
no, because the dark:text-neutral-400 class supersedes the dark:text-red-500 if it's not set.

reneaaron reviewed (approved):
tACK
